src/microbit/Microbit.js

```diff
@@ -1,49 +1,96 @@
-import { createWebBluetoothConnection } from "@microbit/microbit-connection";
+import { ConnectionStatus, createUniversalHexFlashDataSource, createWebBluetoothConnection, createWebUSBConnection } from "@microbit/microbit-connection";
 
 class Microbit {
     constructor() {
-        this.connection = createWebBluetoothConnection();
+        this.bluetooth = createWebBluetoothConnection();
         this.uartDataListener = null;
-        this.connect = this.connection.connect.bind(this.connection);
+        this.bluetoothConnect = this.bluetooth.connect.bind(this.bluetooth);
+
+        this.usb = createWebUSBConnection();
+        this.usbConnect = this.usb.connect.bind(this.usb);
 
         // Initialise micro:bit UART data listener.
         this.uartDataListener = (event) => {
             const decoded = new TextDecoder().decode(event.value);
-            const values = decoded.split(":", 3);
-            const [start, command, arg] = values;
-            if (values.length !== 3 || start !== "c" || isNaN(parseInt(arg))) {
-                throw new Error(`Invalid micro:bit UART message: ${decoded}`);
+            this.triggerCommand(decoded);
+        };
+        this.bluetooth.addEventListener("uartdata", this.uartDataListener);
+
+        // Initialise micro:bit serial data listener.
+        this.serialBuffer = "";
+        this.serialDataListener = (event) => {
+            if (this.bluetooth.status === ConnectionStatus.CONNECTED) {
+                // Listen to UART data listener instead.
+                return
             }
-            switch (command) {
-                case "photo": {
-                    const classIdx = parseInt(arg);
-                    const event = new CustomEvent("record", {
-                        detail: GLOBALS.recording
-                            ? { stop: classIdx }
-                            : { start: classIdx },
-                    });
-                    window.dispatchEvent(event);
-                }
+            const decoded = event.data;
+            this.serialBuffer = this.serialBuffer + decoded;
+            if (this.serialBuffer.endsWith("\n")) {
```

Comment:
Is this guaranteed to be at the end? Or can we have got all of one message and some of the next one?